src/main/java/com/amazon/opendistroforelasticsearch/sql/utils/SQLFunctions.java

```diff
@@ -710,6 +750,96 @@ private String upper(String property, String culture) {
+             * a is parsed as the key (String) of a KVValue (get from paramers.get(0))
+             * and b is parsed as the value (Object) of this KVValue.
+             *
+             * Either a or b could be a column name, literal, or a number:
+             * - if isNumeric is true --> number
+             * - else if this string is single quoted --> literal
+             * - else --> column name
+             */
+            String key = getPropertyOrValue(paramers.get(0).key);
+            String value = getPropertyOrValue(paramers.get(0).value.toString());
+            String condition = key + " == " + value;
+            return ifFunc(condition, expr1, expr2);
+        }
+    }
+
+    private Tuple<String, String> ifFunc(String condition, String expr1, String expr2) {
+        String name = nextId("if");
+        return new Tuple<>(name, "boolean cond = " + condition + ";"
+                + def(name, "cond ? " + expr1 + " : " + expr2));
+    }
+
+    private Tuple<String, String> ifFunc(Boolean condition, String expr1, String expr2) {
+        String name = nextId("if");
+        if (condition) {
+            return new Tuple<>(name, def(name, expr1));
+        } else {
+            return new Tuple<>(name, def(name, expr2));
+        }
+    }
+
+    private Tuple<String, String> ifnull(SQLExpr condition, SQLExpr expr) {
+        String name = nextId("ifnull");
+        if (condition instanceof SQLNullExpr) {
+            return new Tuple<>(name, def(name, expr.toString()));
+        }
+        if (isProperty(condition)) {
+            return new Tuple<>(name, def(name, doc(condition) + ".size()==0 ? " + expr.toString() + " : "
+                    + getPropertyOrValue(condition)));
+        } else {
+            String condStr = Strings.isNullOrEmpty(condition.toString()) ? null : getPropertyOrStringValue(condition);
+            return new Tuple<>(name, def(name, condStr));
+        }
+    }
+
+    private Tuple<String, String> isnull(SQLExpr expr) {
+        String name = nextId("isnull");
+        if (expr instanceof SQLNullExpr) {
+            return new Tuple<>(name, def(name, "1"));
+        }
+        if (isProperty(expr)) {
+            return new Tuple<>(name, def(name, doc(expr) + ".size()==0 ? 1 : 0"));
+        }
+        // cases that return 1:
+        // expr is null || expr is math func but tends to throw "divided by zero" arithmetic exception
+        String resultStr = "0";
+        if (Strings.isNullOrEmpty(expr.toString())) {
+            resultStr = "1";
+        }
+        if (expr instanceof SQLCharExpr && this.generatedIds.size() > 1) {
+            // the expr is a math expression
+            String mathExpr = ((SQLCharExpr) expr).getText();
+            return new Tuple<>(name, StringUtils.format(
+                    "try {%s;} "
+                            + "catch(Exception e) "
+                            + "{def %s; if(e instanceof ArithmeticException){%s=1;} else{%s=0;} return %s;} "
```

Comment:
maybe a little better?

```
try { } 
catch (ArithmeticException e) { return 1; }
return 0;
```

Reply:
There is a post-processing of the functionStr right after the function method switch: 
`String returnCommand = ";return " + generatedFieldName + ";";`
That's why I made it that complicated.

Reply:
But I will make it catch arithmetic exception rather than generic exception

Reply:
I see. Thanks.